components/style/values/specified/image.rs

```diff
@@ -0,0 +1,402 @@
+        } else {
+            Ok(AngleOrCorner::Angle(Angle(PI)))
+        }
+    }
+}
+
+impl ComputedValueAsSpecified for AngleOrCorner {}
+
+/// Specified values for one color stop in a linear gradient.
+/// https://drafts.csswg.org/css-images/#typedef-color-stop-list
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ColorStop {
+    /// The color of this stop.
+    pub color: CSSColor,
+
+    /// The position of this stop. If not specified, this stop is placed halfway between the
+    /// point that precedes it and the point that follows it.
+    pub position: Option<LengthOrPercentage>,
+}
+
+impl ToCss for ColorStop {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        try!(self.color.to_css(dest));
+        if let Some(position) = self.position {
+            try!(dest.write_str(" "));
+            try!(position.to_css(dest));
+        }
+        Ok(())
+    }
+}
+
+define_css_keyword_enum!(HorizontalDirection: "left" => Left, "right" => Right);
+define_css_keyword_enum!(VerticalDirection: "top" => Top, "bottom" => Bottom);
+
+impl Parse for ColorStop {
+    fn parse(input: &mut Parser) -> Result<Self, ()> {
+        Ok(ColorStop {
+            color: try!(CSSColor::parse(input)),
+            position: input.try(LengthOrPercentage::parse).ok(),
+        })
+    }
+}
+
+/// Determines whether the gradient's ending shape is a circle or an ellipse.
+/// If <shape> is omitted, the ending shape defaults to a circle
+/// if the <size> is a single <length>, and to an ellipse otherwise.
+/// https://drafts.csswg.org/css-images/#valdef-radial-gradient-ending-shape
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum EndingShape {
+    Circle(LengthOrKeyword),
+    Ellipse(LengthOrPercentageOrKeyword),
+}
+
+impl Parse for EndingShape {
+    fn parse(input: &mut Parser) -> Result<Self, ()> {
+        match_ignore_ascii_case! { try!(input.expect_ident()),
+            "circle" => {
+                let position = input.try(LengthOrKeyword::parse).unwrap_or(
+                    LengthOrKeyword::Keyword(SizeKeyword::FarthestSide));
+                Ok(EndingShape::Circle(position))
+            },
+            "ellipse" => {
+              let length = if let Ok(len) = input.try(LengthOrPercentageOrKeyword::parse) {
+                    len
+                } else {
+                    (LengthOrPercentageOrKeyword::Keyword(SizeKeyword::FarthestSide))
+                };
+                Ok(EndingShape::Ellipse(length))
+            },
+            _ => {
+                // If two <length> is present, it defaults to ellipse, otherwise defaults to circle.
+                if let Ok(length) = LengthOrPercentageOrKeyword::parse(input) {
+                    if let LengthOrPercentageOrKeyword::Keyword(keyword) = length {
+                        Ok(EndingShape::Circle(LengthOrKeyword::Keyword(keyword)))
```

Comment:
nit: leave a comment here saying "a single keyword is valid for both ellipse and circle, but we default to circle. The grammar for ending shapes for circle and ellipse have overlap so we cannot simply try to parse as circle first"